File: desktop/windows/Nexus.Desktop/Bridge/DesktopPersistentStateStore.cs
using System.IO;
using System.Text.Json;
using Nexus.Desktop.Sidecar;

namespace Nexus.Desktop.Bridge;

internal static class DesktopPersistentStateStore
{
    private static readonly object SyncRoot = new();
    private static readonly string StatePath = Path.Combine(DesktopPaths.ConfigDirectory, "desktop-state.json");

    public static string? Get(string key)
    {
        string normalizedKey = NormalizeKey(key);
        lock (SyncRoot)
        {
            return ReadAll().TryGetValue(normalizedKey, out string? value) ? value : null;
        }
    }

    public static void Set(string key, string value)
    {
        string normalizedKey = NormalizeKey(key);
        lock (SyncRoot)
        {
            Dictionary<string, string> values = ReadAll();
            values[normalizedKey] = value;
            WriteAll(values);
        }
    }

    public static void Remove(string key)
    {
        string normalizedKey = NormalizeKey(key);
        lock (SyncRoot)
        {
            Dictionary<string, string> values = ReadAll();
            values.Remove(normalizedKey);
            WriteAll(values);
        }
    }

    private static string NormalizeKey(string key)
    {
        string normalized = key.Trim();
        if (string.IsNullOrWhiteSpace(normalized) ||
            normalized.Length > 128 ||
            normalized.Any(character => !(char.IsLetterOrDigit(character) || character is '.' or '_' or '-')))
        {
            throw new ArgumentException("Persistent state key is invalid.");
        }

        return normalized;
    }

    private static Dictionary<string, string> ReadAll()
    {
        try
        {
            if (!File.Exists(StatePath))
            {
                return new Dictionary<string, string>();
            }

            string raw = File.ReadAllText(StatePath);
            Dictionary<string, string>? values = JsonSerializer.Deserialize<Dictionary<string, string>>(raw);
            return values ?? new Dictionary<string, string>();
        }
        catch
        {
            return new Dictionary<string, string>();
        }
    }

    private static void WriteAll(Dictionary<string, string> values)
    {
        Directory.CreateDirectory(DesktopPaths.ConfigDirectory);
        string tempPath = $"{StatePath}.tmp";
        File.WriteAllText(tempPath, JsonSerializer.Serialize(values));
        if (File.Exists(StatePath))
        {
            File.Replace(tempPath, StatePath, null);
            return;
        }

        File.Move(tempPath, StatePath);
    }
}

File: desktop/macos/Sources/NexusDesktop/Bridge/DesktopPersistentStateStore.swift
import Foundation

enum DesktopPersistentStateStore {
  private static let queue = DispatchQueue(label: "com.leemysw.nexus.desktop-persistent-state")
  private static var stateURL: URL {
    DesktopPaths.configDirectory.appendingPathComponent("desktop-state.json")
  }

  static func get(_ key: String) throws -> String? {
    let normalizedKey = try normalizeKey(key)
    return try queue.sync {
      let values = try readAll()
      return values[normalizedKey]
    }
  }

  static func set(_ value: String, forKey key: String) throws {
    let normalizedKey = try normalizeKey(key)
    try queue.sync {
      var values = try readAll()
      values[normalizedKey] = value
      try writeAll(values)
    }
  }

  static func remove(_ key: String) throws {
    let normalizedKey = try normalizeKey(key)
    try queue.sync {
      var values = try readAll()
      values.removeValue(forKey: normalizedKey)
      try writeAll(values)
    }
  }

  private static func normalizeKey(_ key: String) throws -> String {
    let normalized = key.trimmingCharacters(in: .whitespacesAndNewlines)
    let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-")
    guard !normalized.isEmpty,
          normalized.count <= 128,
          normalized.rangeOfCharacter(from: allowed.inverted) == nil else {
      throw DesktopPersistentStateError.invalidKey
    }
    return normalized
  }

  private static func readAll() throws -> [String: String] {
    let fileManager = FileManager.default
    guard fileManager.fileExists(atPath: stateURL.path) else {
      return [:]
    }
    do {
      let data = try Data(contentsOf: stateURL)
      let raw = try JSONSerialization.jsonObject(with: data) as? [String: String]
      return raw ?? [:]
    } catch {
      return [:]
    }
  }

  private static func writeAll(_ values: [String: String]) throws {
    let fileManager = FileManager.default
    try fileManager.createDirectory(at: DesktopPaths.configDirectory, withIntermediateDirectories: true)
    let data = try JSONSerialization.data(withJSONObject: values, options: [.sortedKeys])
    let tempURL = stateURL.appendingPathExtension("tmp")
    try data.write(to: tempURL, options: [.atomic])
    if fileManager.fileExists(atPath: stateURL.path) {
      _ = try fileManager.replaceItemAt(stateURL, withItemAt: tempURL)
    } else {
      try fileManager.moveItem(at: tempURL, to: stateURL)
    }
  }
}

private enum DesktopPersistentStateError: LocalizedError {
  case invalidKey

  var errorDescription: String? {
    "Persistent state key is invalid."
  }
}
